crypto/engine/eng_cnf.c

```diff
@@ -76,19 +76,34 @@ static int int_engine_configure(const char *name, const char *value, const CONF
         else if (strcmp(ctrlname, "soft_load") == 0)
             soft = 1;
         /* Load a dynamic ENGINE */
-        else if (strcmp(ctrlname, "dynamic_path") == 0) {
-            e = ENGINE_by_id("dynamic");
-            if (!e)
-                goto err;
-            if (!ENGINE_ctrl_cmd_string(e, "SO_PATH", ctrlvalue, 0))
-                goto err;
-            if (!ENGINE_ctrl_cmd_string(e, "LIST_ADD", "2", 0))
-                goto err;
-            if (!ENGINE_ctrl_cmd_string(e, "LOAD", NULL, 0))
-                goto err;
-        }
+        else if (strcmp(ctrlname, "dynamic_path") == 0)
+            dynamic_path = ctrlvalue;
         /* ... add other pseudos here ... */
         else {
+            /* Load a dynamic ENGINE now that we are sure about it's engine_id */
+            if (dynamic_path) {
+                /*
+                 * Check if the engine exists before we try to load it.
+                 * This solves various errors caused by trying to load
+                 * the same engine more than once.
+                 */
+                e = ENGINE_by_id(name);
```

Comment:
in the case that the engine is not in the list, this does more than what your comment says:
this will try to load an engine with the given name in the engines directory.

Reply:
Is there any way to query if a given engine (by engine_id) is already in the engine list? And if not, can I add one so it can be used for this purpose?

Reply:
The problem is your patch creates a race condition between the query
for the duplicate name, and the `ENGINE_ctrl_cmd_string(e, "LOAD", NULL, 0)`
the engine with the same name may be loaded by a different thread, in between.

Reply:
I understand your concern, and am working on a new revision for this commit in which the check is part of the "load" command in eng_dyn.c.

That said, the documentation for ENGINE_by_id() is: "Retrieve an engine from the list by its unique "id" value.", while in practice it also tries to load an engine with the given name from the engines directory. When checking for the existence of an engine, should I implement a new ENGINE_search() method?

As for the race condition, seeing that I will move the check to be inside the "load" cmd, it will resolve the issue of a 3rd party program loading the config after OpenSSL loaded it on its own, and it will drastically reduce the window for a race condition. The fix itself meant to be a robust workaround to the issue of crashing when loading the engines twice. That said, engines shouldn't get loaded twice in the first place, so if we do have a race condition that causes this scenario I think it is something we could live with.

Reply:
yeah, something like an engine_search, with lowercase so it's internal.
one other concern that I have is that it may well be possible to use more than one dynamic_path
to load several engines, at once.

Reply:
I didn't see any configuration file with such a setup, and anyway the configuration parser will only save the last line of a given key. I just tried to add multiple such lines to the file, and added a debug print and the for loop only showed the last "dynamic_path" line in the file. So we are covered.

I'll update my commit now, and upload it.

Reply:
ah, good.  thanks for double checking.

Reply:
I've looked for where engine "LOAD" is used, and found another place:
https://github.com/openssl/openssl/blob/af16097febcd4fa31cd5fcd05ad09cf8b53659ea/apps/lib/engine.c#L32-L37
and indeed it is easy to create a memory leak there:
`../util/shlib_wrap.sh ./openssl s_client -engine $PWD/../engines/ossltest.so  -engine $PWD/../engines/ossltest.so`

Reply:
Would it work to use the address of the bind function as an ID when the ENGINE_ctrl_cmd "ID" is not used ?